compiler/src/iree/compiler/Dialect/LinalgExt/IR/AggregatedOpInterfaceImpl.cpp

```diff
@@ -1043,4 +1044,190 @@ FailureOr<SmallVector<Value>> CustomOp::decomposeOperation(OpBuilder &builder) {
+    }
+    rewriter.replaceAllUsesWith(linalgOp.getResult(resultNumber),
+                                newOp.getResult(0));
+
+    results.emplace_back(newOp);
+  }
+
+  return results;
+}
+
+FailureOr<SmallVector<Value>> ExpReductionOp::decomposeOperation(OpBuilder &b) {
+  Location loc = getLoc();
+  IRRewriter rewriter(b);
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(*this);
+
+  // Split the op into:
+  // curr_max = max(s, old_max)
+  // ex = e^{x - curr_max}
+  // norm = e^{curr_max - old_max}
+  // for each outs in exp_reduction:
+  //     norm_outs = outs * norm
+  // linalg.generic ins(ex, ...) outs(norm_outs)
+
+  const int reducingOpIndex = getReducingOpIndex();
+  OpOperand *sValue = getDpsInputOperand(reducingOpIndex);
+  OpOperand *prevMax = getDpsInitOperand(reducingOpIndex);
+  AffineMap normValMap = getMatchingIndexingMap(sValue);
+  AffineMap prevMaxMap = getMatchingIndexingMap(prevMax);
+
+  // curr_max = max(sValue, prev_max)
+  Value currMax = reduce<arith::MaximumFOp>(
+      rewriter, loc, normValMap, prevMaxMap, sValue->get(), prevMax->get());
+  // ex = e^{sValue - curr_max}
+  Value ex = computeSubAndExp2(rewriter, loc, prevMaxMap, normValMap, currMax,
+                               sValue->get());
+  // norm = e^(prev_max - curr_max)
+  Value norm = computeSubAndExp2(rewriter, loc, prevMaxMap, prevMaxMap, currMax,
+                                 prevMax->get());
+
+  SmallVector<Value> inputs = getDpsInputs();
+  SmallVector<Value> normOuts(getNumDpsInits());
+  inputs[reducingOpIndex] = ex;
+  normOuts[reducingOpIndex] = currMax;
+  for (int64_t oldIndex : getExpReducedOperands()) {
+    OpOperand *oldOut = getDpsInitOperand(oldIndex);
+    AffineMap oldOutMap = getMatchingIndexingMap(oldOut);
+    Value normOut = elementwiseValueInPlace<arith::MulFOp>(
+        rewriter, loc, oldOutMap, prevMaxMap, oldOut->get(), norm);
+    normOuts[oldIndex] = normOut;
+  }
+
+  auto expRedGeneric = linalg::GenericOp::create(
+      rewriter, loc, TypeRange(normOuts), inputs, normOuts,
+      getIndexingMapsArray(), getLoopIteratorTypes());
+  expRedGeneric->setDiscardableAttrs(
+      getOperation()->getDiscardableAttrDictionary());
+
+  IRMapping mapper;
+  getBodyRegion().cloneInto(&expRedGeneric.getBodyRegion(), mapper);
+  rewriter.setInsertionPointAfter(expRedGeneric.getBody()->getTerminator());
+  auto yieldOp =
+      cast<IREE::LinalgExt::YieldOp>(expRedGeneric.getBody()->getTerminator());
+  rewriter.replaceOpWithNewOp<linalg::YieldOp>(yieldOp, yieldOp.getOperands());
+
+  FailureOr<SmallVector<linalg::GenericOp>> decomposedResults =
+      decomposeMultipleResults(expRedGeneric, rewriter);
+  if (failed(decomposedResults)) {
+    return failure();
+  }
+  return llvm::map_to_vector(decomposedResults.value(),
+                             [](linalg::GenericOp op) -> Value {
+                               assert(op->getNumResults() == 1);
```

Comment:
The assertion is redundant to me, because `decomposeMultipleResults` should always give you the expected generic op.